File: server/app/services/perenual_service.py
import requests
import os
from dotenv import load_dotenv

PERENUAL_API_KEY = os.getenv("PERENUAL_API_KEY")
PERENUAL_BASE_URL = "https://perenual.com/api/species/details"

def fetch_plant_from_perenual(plant_id: str):
    """Fetch plant details from Perenual API and adapt for Pydantic validation"""
    url = f"{PERENUAL_BASE_URL}/{plant_id}?key={PERENUAL_API_KEY}"
    response = requests.get(url)
    
    if response.status_code == 200:
        plant_data = response.json()
        
        # Rename 'id' to 'plant_id' to match PlantModel
        plant_data["plant_id"] = str(plant_data.pop("id", plant_id))  # Default to provided ID if missing

        return plant_data  # Return the modified dictionary
    return None

File: server/services/perenualService.js
import axios from "axios";
import dotenv from "dotenv";

dotenv.config();

const PERENUAL_API_KEY = process.env.VITE_APP_PERENUAL_API_KEY;
const PERENUAL_BASE_URL = "https://perenual.com/api";

/**
 * Fetch plant details from Perenual API
 */
export const fetchPlantFromPerenual = async (plantName) => {
    try {
        console.log(`🔍 [Perenual Service] Fetching plant data for: ${plantName}`);

        // ✅ If Vision returned "Flower", try querying with a more specific term
        if (plantName.toLowerCase() === "flower" || plantName.toLowerCase() === "plant") {
            console.log("⚠️ [Perenual Service] Generic term detected, refining search...");
            plantName = "Rose"; // ✅ Use a more specific default plant name
        }

        const response = await axios.get(`${PERENUAL_BASE_URL}/species-list`, {
            params: { key: PERENUAL_API_KEY, q: plantName }
        });

        if (response.status === 200 && response.data.data.length > 0) {
            console.log(`✅ [Perenual Service] Plant Data Found:`, response.data.data[0]);
            return response.data.data[0]; // ✅ Return the first matching plant
        } else {
            console.log(`⚠️ [Perenual Service] No results found for: ${plantName}`);
            return null;
        }
    } catch (error) {
        console.error(`❌ [Perenual Service] Failed to fetch plant from Perenual API:`, error.message);
        return null;
    }
};


/**
 * Analyze plant health based on Google Vision results
 * @param {Array} labels - Labels returned from Google Vision API
 * @returns {Object} - Plant health analysis
 */
export const analyzePlantHealth = (labels) => {
    const diseasePatterns = ['spot', 'lesion', 'damage', 'wilt', 'yellowing'];

    const detectedIssues = labels.filter(label => 
        diseasePatterns.some(pattern => label.description.toLowerCase().includes(pattern))
    );

    return {
        isHealthy: detectedIssues.length === 0,
        details: detectedIssues.length > 0 
            ? `Detected potential issues: ${detectedIssues.map(l => l.description).join(", ")}` 
            : 'No visible health issues detected'
    };
};
